### PR TITLE
Add missing groups fitted from thermo libraries

### PR DESCRIPTION
**Pull request by hwpang**
# Changes Made

Thermo entries from the following thermo libraries were used to extract and fit missing groups.

```
libraries = ['Klippenstein_Glarborg2016', 'BurkeH2O2', 'thermo_DFT_CCSDTF12_BAC', 'DFT_QCI_thermo',
                         'primaryThermoLibrary', 'primaryNS', 'NitrogenCurran', 'NOx2018', 'FFCM1(-)',
                         'SulfurLibrary', 'SulfurGlarborgH2S', 'SABIC_aromatics']
```

128 groups were identified as missing and 134 species were used to fit the missing groups using ridge regression.

Uncertainties were also calculated for each fitted groups.

The results with the new fitted groups are shown below.

image

image

# Review Tips

I used this ipython notebook to identify missing groups, fit group values, calculate uncertainties, generate group objects and write new group to database. It might be easier to review the notebook: https://www.dropbox.com/s/270c4easr46b18o/GAV%20from%20thermo%20libraries%20-%20hwpang%20-%2020210119.ipynb?dl=0

### CODE REVIEW

**Comment by hwpang** [comment]
Looks like the database test failed. I will check out that soon.

**Comment by hwpang** [comment]
In the new push, I slightly change the way the code automatically writes group structures, so that the sample molecule made from each group could match to their node.

Here's my strategy to automatically create groups:
1) if the labeled atom is not in a cycle, use its neighbor to define group
1) if the labeled atom is in a cycle, use the ring member and its neighbor to define group
1) if the sampled molecule still fails to match to its own node, create a specific group using all non-H atom for specification, but change the end carbon to a generic carbon for generalization.

The new push identified 143 missing group and fitted them with 144 species using Ridge regression with 0.001 as hyperparameter. While not having enough data to autotune the hyperparameter, 0.001 has the best among my manual trials (0.001, 0.01, and 0.1).

The new push has results shown below.

image


image


image


image

**Comment by hwpang** [comment]
In the newest push, I slightly change the way the code automatically writes group structures to make sure group names are distinguishable.

The new push identified 141 missing group and fitted them with 144 species using Ridge regression with 0.001 as hyperparameter. 

Per suggestion by @kspieks offline, I use 1-fold cross validation to validate the fitted groups, as shown below. Trials show that the validation error from 1-fold cross validation is not sensitive to the hyperparameter I use.

image

image

image

**Comment by hwpang** [comment]
In the newest push, I slightly change the way the code automatically create label for groups to make sure group names are distinguishable.

The new push identified 141 missing group and fitted them with 144 species using Ridge regression with 0.01 as hyperparameter. Per suggestion by @kspieks offline, I use the leave-one-out cross validation to validate the fitted groups, as shown below.

image
image

I tuned the hyperparameter using the leave-one-out cross validation. A hyperparameter of 0.1 has the lowest validation RMSE, but 0.01 and smaller has the smaller validation MAE, as shown below. I chose 0.01 as hyperparameter.

image
image

The parity plot with uncertainties and the training error are shown below.
image
image

**Comment by kspieks** [comment]
Thanks for the PR and great addition to the database! Thanks for working on leave one out xval and including hyperparameter sweeps for RMSE and MAE. I think the new histograms and parity plots look good. The outliers on the new parity plots seem to have even smaller errors than the parity plots pushed on Jan 22. I appreciate you including the notebook and discussing the fitting offline :) @amarkpayne can you look at this when you get a chance? I think it's ready to go, but would appreciate your experienced perspective to make sure I didn't miss anything :)

**Review by kspieks — APPROVED** [review]

**Comment by davidfarinajr** [comment]
Hey guys!  This looks great, thanks for the contribution!  I really like this approach for adding new thermo groups to the database.  I was wondering if you've seen this PR https://github.com/ReactionMechanismGenerator/RMG-database/pull/434 which adds new nitrogen and halogen thermo groups.  Do you think you could test the thermo groups in this PR against the same g4mp2 test set (https://pubs.rsc.org/en/content/articlehtml/2019/sc/c9sc02834j)?  If you're interested, I could send you a csv with the g4mp2 enthalpies.  Also, since RMG's libraries don't contain much nitrogen thermo, I added a `CHON_G4` thermo library here (https://github.com/ReactionMechanismGenerator/RMG-database/pull/434/commits/7bfed9d768312490f9c53728f1e980483ffe64de) and used the nitrogen species in this library to derive the nitrogen thermo groups in the PR.  Maybe we could merge that library into this branch and derive more nitrogen groups, if need be, using this missing group approach?

**Comment by hwpang** [comment]
> Hey guys! This looks great, thanks for the contribution! I really like this approach for adding new thermo groups to the database. I was wondering if you've seen this PR #434 which adds new nitrogen and halogen thermo groups. Do you think you could test the thermo groups in this PR against the same g4mp2 test set (https://pubs.rsc.org/en/content/articlehtml/2019/sc/c9sc02834j)? If you're interested, I could send you a csv with the g4mp2 enthalpies. Also, since RMG's libraries don't contain much nitrogen thermo, I added a `CHON_G4` thermo library here ([7bfed9d](https://github.com/ReactionMechanismGenerator/RMG-database/commit/7bfed9d768312490f9c53728f1e980483ffe64de)) and used the nitrogen species in this library to derive the nitrogen thermo groups in the PR. Maybe we could merge that library into this branch and derive more nitrogen groups, if need be, using this missing group approach?

Thanks for the positive comment and suggestion! I am definitely interested in testing these thermo groups again the g4mp2 dataset you mentioned (my email is hwpang@mit.edu :D ) And I think it's an excellent idea to merge the PR of CHON_G4 thermo library and derive more nitrogen groups using this missing group approach! I'm very excited to test this approach with more data and add more nitrogen groups to RMG database!

**Comment by mjohnson541** [comment]
@hwpang can you rebase?

**Comment by hwpang** [comment]
I rebased the branch and refitted the groups based on the newest commit on branch `main` using the same notebook. I also added the notebook to script for future reference.

**Comment by hwpang** [comment]
The results from the newest push are shown below.  72 species were used to fit 57 groups. The loss in number of missing groups in the new push is because many missing groups fitted in the previous push were addressed by David's PR merged in previously.

Before
image
After
image
Before
image
After
image

Hyperparameter tuning
image

image


1-cross validation
image

image

**Comment by hwpang** [comment]
Newest push addresses some comments from @kspieks offline, including to explain the role of the special list. 

`special_list` contains groups that always appear with other groups such that these groups are not independent. 
For example, O2d-CO always appear with CO-O2d, and S2d-CS always appear with CS-S2d.
In the past, RMG developers chose to set these groups to be zeros, so we don't generate new groups on them.
This list needs to be updated when the relevant groups are changed.

The new push generates 57 group from 79 species.

Before
image

After
image


Before
image


After
image

Hyperparameter tuning
image

image


one-fold cross-validation
image

image

**Review by xiaoruiDong — COMMENTED** [review]
@hwpang Thanks for the contribution! The visualization is helpful and the results look promising. Just a minor comment about the notebook. Can you please further clean up the notebook, increasing its written quality and readability. Several aspects you may consider:
## Cleanup
- Remove the lines that are commented.
- Separate `user input`, `function definition`, `function calling`
## Guidance for users
- Add more instructions for each block (E.g., the block `check if there's bad fit`, what is this block for? what output should a user expect?) Add section numbers to indicate the sequence (1., 1.1, 2.1, etc.)
- Standardize the output (following the same idea as `add more instructions`, e.g., print spc_num and grp_num in a structured string other than two separate cells.
## Instruction for developers
- If possible, add more comments in function definitions, especially at the place when multiple layers of for loops and if else are stacked. Add some comments to explain why something needs to be treated specially
- If possible, make sure that each function has a docstring to provide guidance for arguments.

**Comment by hwpang** [comment]
@xiaoruiDong Thanks for the comment! I think the new push addresses all the mentioned issues. I also slightly improve the `make_group` algorithm with the new recursive functions. And I changed the code to use RidgeCV for hyperparameter tuning and RMSE vs. hyperparameter plots, instead of my previous manual ones.

The new push fits 33 of new groups with 40 of species. The optimal hyperparameter is 1e-06, found by one-fold cross-validation using RidgeCV. The results from this push are shown below.

Hyperparameter tuning

image

image

H298 before
image

H298 after
image

S298 before
image

S298 after
image

1-fold cross validation
image


image

**Comment by hwpang** [comment]
The new push makes minor changes on the comments in the notebook

**Review by xiaoruiDong — COMMENTED** [review]
The new notebook looks pretty awesome! Please go ahead and remove the temporary commit!

**Comment by hwpang** [comment]
@xiaoruiDong Thanks for the review! The new push removes the temporary commit for twin PR.
